### DISCUSSION TITLE
터놓고 인프라 구성 논의

### DISCUSSION THREAD

**Discussion by soominsohn** [woowacourse-teams/2022-ternoko#216] [General]
터놓고 인프라 구성 제안 드립니다!

인프라 구성도 일단 요구사항을 만족하는 돌아가는 인프라를 구성한 뒤, 점진적인 리팩토링을 통해 완성하고 싶습니다.

제안하는 터놓고 인프라의 최종 목표는 다음과 같습니다.

![image](https://user-images.githubusercontent.com/83059234/184270757-d9ec2787-612a-4025-b1f4-bf24a3fe1333.png)

**최종 목표**

- 젠킨스 인스턴스 2개
    - DEV용 젠킨스
        - 파이프라인 도입
        - develop 브랜치 빌드
        - 라벨로 frontend, backend 빌드 분리
    - PROD용 젠킨스
        - 파이프라인 도입
        - main 브랜치 빌드
        - 라벨로 frontend, backend 빌드 분리
- 백엔드 인프라 구성
    - DEV (dev.ternoko.site)
        - WS (nginx)
        - WAS(Spring + DB)
    - PROD (api.ternoko.site)
        - WS (nginx)
        - WAS(Spring)
        - DB
- 프론트엔드 인프라 구성
    - DEV
        - WS + WAS (nginx + React)
    - PROD
        - WS + WAS(nginx + React)

- 총 필요 인스턴스: **9개**

1. 젠킨스 관련 논의 사항
    - 제가 제안하는 터놓고의 CI/CD를 위한 젠킨스 활용방안은 다음과 같습니다.
        - **젠킨스를 DEV용, PROD용으로 나눈다.**
            - DEV와 PROD 젠킨스를 하나로 합칠 시, 부하가 한군데로 가 PROD 릴리즈 시에 문제가 생길 가능성이 있어보입니다. 따라서 분리하는 것이 좋을 것 같습니다.
        - **파이프라인 도입 후, 프론트와 백엔드 CI/CD과정을 한 젠킨스에서 돌린다.**
            - 만일 프론트와 백엔드 CI/CD를 개별의 젠킨스에서 돌릴 경우 DEV, PROD 분리 시 총 4개의 젠킨스 인스턴스가 필요한데, 과연 그만큼의 인스턴스를 써도 되는지… 하는 생각이 들었습니다.
            - 또한 현재 디스크 용량 조절이 가능한 것으로 보아 파이프라인을 도입해 프론트와 백엔드를 같은 젠킨스 서버에 두어도 될 것 같습니다.
            - 파이프라인 도입 전에는, 일단 돌아가는 인프라를 만들기 위해 프론트엔드/백엔드 젠킨스 인스턴스를 따로 생성할 예정입니다.

1. 프론트엔드 PROD, DEV 분리
    - 프론트엔드 또한 PROD와 DEV 분리가 필요해 보입니다.
        - PROD - ternoko.site
        - DEV - front.ternoko.site
        - 프론트용 dev 도메인 새로 생성
        - front.ternoko.site 처럼 서브도메인 생성해서 분리하는 것이 어떨까요?
    

제안하는 최종 인프라 구성은 다음과 같고,

일단 요구사항 만족을 위해

백엔드 개발서버와 운영서버 분리를 먼저 진행할 예정입니다. 

혹시 보시고 궁금한 사항 있으시면 자유롭게 답변이나 슬랙, 찾아와주세용!

@lokba @her0807 @Juhyung990122 @HyeonbinSa @dongho108 @sanaandmomo

**Comment by HyeonbinSa** [comment]
우선, 제가 듣기로 13개의 인스턴스를 사용하는 다른 조도 있는거 같아 9개 정도면 충분히 사용할 수 있다고 생각합니다.(특히, 운영 서버와 개발 서버를 분리해야하기때문에)
그래서 작성해주신 인프라 구성도를 보고 전반적으로 동의합니다!

하지만 의견을 하나 제시하자면 지금 NginX가 백엔드/프론트엔드에 각각 하나씩 존재하는데 하나의 NginX를 통해 백, 프론트에 뿌려주는 동작을 구현할 수 있다는 이야기를 들어 추후에 고민해보면 좋을 것 같습니다. 현재 구성도 좋아 당장 바꾸자는 것은 아니고 나중에 한번 고민해보면 좋을 것 같은 구성이라 의견만 드립니다.

**Reply by soominsohn** [reply]
하나의 NginX를 통해 백, 프론트에 뿌려주는 동작에 관해서 제 의견 말씀드립니다.
- nginx하나로 백, 프론트의 모든 요청을 받게 된다면, 모든 부하가 한대의 nginx서버로 들어갑니다.
- 또한 공통 nginx서버를 두게 된다면 작동하지 않을 때 프론트 페이지로 접근 불가능한 문제가 생깁니다. nginx를 분리한다면 서버장애가 나도 프론트서버에는 이를 표시할 에러페이지를 보여 줄 수 있습니다. 
그래서 저는 nginx 두개를 두는 것을 제안드립니다.

**Reply by HyeonbinSa** [reply]
답변 감사해요. 앤지!! 서버 장애를 대비한 분리였군요. 
말씀해주신 내용을 생각해보니 앤지가 작성해준 인프라 구성이 옳은 것 같습니다! 
다시 한번 설명해주신 점 감사드려요

**Comment by dongho108** [comment]
develop 서버도 DB 분리해야 할 것 같습니다.
prod 서버와 동일하지만 DB만 다르게 구성하여 develop에서 돌아가면 prod에서도 돌아간다고 확신할 수 있어야할 것 같아용.
그리고 initializer도 develop, prod 에 넣어야할지 고민해야할 것 같아요

**Comment by Juhyung990122** [comment]
안녕하세요 앤지! 제안 주신 인프라 구성 잘 보았습니다:) 열심히 구상해주셨더라구요!
전반적으로 동의하며 논의사항에 대한 제 생각 남깁니다~!

# 젠킨스를 DEV용, PROD용으로 나눈다.
현재 다른 팀중에  모든 젠킨스를 하나로 관리하는(쩐다..) 팀이 있다고 하는데요. 
들어보니 빌드시 메모리 할당을 해놔서 가능한 것 같아요. 할아버지가 뻗지는 않는데 한번 할때마다 6~10분정도 걸린다고 하더라구요..ㅎㅎ
이에 저희거가 뻗는 이유도 캐시,로그가 차지하는 메모리가 아니라 빌드로 인한 메모리가 아닐까 추측중입니다.
그래서 제안 주신것처럼 전부 분리하는 것도 좋으나 
저는 **프론트엔드 백엔드 두개로만 나누고, 수동 메모리할당을 먹이는것**도 좋겠다는 생각입니다.
이유는 다음과 같습니다.

## 빌드 타이밍 조절 가능
앞서 말씀드린대로 캐시, 로그가 먹는 메모리가 원인이 아니라 빌드시 메모리가 문제라면 
수동으로 빌드메모리를 할당한 뒤 DEV와 PROD를 한 젠킨스에서 보아도 될 것 같습니다. 
일단 DEV와 PROD가 동시에 빌드를 하는 경우가 많지 않을 것 같고, 
사실 이 빌드 타이밍은 저희 팀 깃허브 머지기준이기 때문에 팀 내에서 컨트롤할 수 있는 부분이여서요.
지금보다 빌드속도가 느려지긴 하겠지만 4개도 안뻗고 돌리니까.. 2개면.. 그래도 나쁘지않을 것 같아요.
다만 프엔 백엔은 작업속도나 머지타이밍이 달라서 DEV와 PROD를 분리하기보다는 FE,BE을 분리했으면 좋겠다는 생각입니다. 

## 젠킨스 관리포인트를 줄일 수 있음
젠킨스 4개.. 생각만해도 빡세군요.. 인스턴스 생성 갯수 제한은 없으나 관리측면에서는 아무래도 적을수록 좋다는 생각입니다.
그래서 2개로 간다면 간단하고 좋을 것 같아요. 4개로 가게된다면 마스터노드 서버를 하나 두고 나머지 4개 서버를 슬레이브로 등록해서 마스터에서 일괄관리해도 좋을 듯 합니다. 

# 파이프라인 도입 후, 프론트와 백엔드 CI/CD과정을 한 젠킨스에서 돌린다.
동의합니다. 부하상황을 봐서 하나로 합치는게 제일 베스트일 것 같아요.
디스크 조절이 가능하긴 하나... 저희 그거...공적인 공간이라 말을 아끼겠습니다. 
근데 사실 괜찮을지 걱정돼요. 방학끝나고 서버 없어져있음 어쩌죠..
아무튼 도입 이전에는 두개로, 도입 후에는 상황에 따라 하나로 합치는건 동의합니다!

# 그래서 이열음 넌 뭘 어쩌고 싶냐 
그래서 저는 대충 이런 구성도를 생각해보았는데요.

![IMG_1CDB9C63281E-1](https://user-images.githubusercontent.com/43205258/184518725-ae11deb5-16ab-472d-a40c-d58371fa6160.jpeg)
-> 이미지 내에 화살표 하나가 누락됐네요 BE Jenkins가 DEV랑 PROD 둘 다 봅니다!

앤지랑 전반적으로 비슷하나 유사한 환경을 공유하는 경우가 많아서 docker 사용도 고려하여 구성해보았습니다.
이러면 한번 뜬 이미지로 사용 가능해서 구성하는데 시간을 좀 덜 수 있을 것 같네요
한번 보시고 의견 말씀해주셔요ㅎㅎ 고생많으셨습니다. 

+) 고민하느라 답변이 늦어진 점 죄송합니당ㅠㅎ

**Comment by soominsohn** [comment]
[09/13 구성도] 
![image](https://user-images.githubusercontent.com/83059234/189808425-6013d279-9a8f-4db7-a2d2-56b5e41a7946.png)
